hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/StreamKafkaP.java

```diff
@@ -71,28 +75,58 @@ private StreamKafkaP(String[] topicIds, Properties properties) {
     @Override
     protected void init(@Nonnull Context context) throws Exception {
         jobFuture = context.jobFuture();
+        consumer = new KafkaConsumer<>(properties);
+        consumer.subscribe(Arrays.asList(topicIds));
     }
 
     @Override
     public boolean complete() {
-        try (KafkaConsumer<?, ?> consumer = new KafkaConsumer<>(properties)) {
-            consumer.subscribe(Arrays.asList(topicIds));
-
-            while (!jobFuture.isDone()) {
-                ConsumerRecords<?, ?> records = consumer.poll(POLL_TIMEOUT_MS);
-
-                for (ConsumerRecord<?, ?> r : records) {
-                    emit(entry(r.key(), r.value()));
-                }
-                consumer.commitSync();
+        if (pendingEntry != null) {
+            // If successfully emit the pending entry, resume the consumer
+            // else do an empty poll for heartbeat
+            if (tryEmit(pendingEntry)) {
+                pendingEntry = null;
+                resume();
+            } else {
+                emptyPoll();
+                return jobFuture.isDone();
+            }
+        }
+        // poll for records if iterator is null
+        if (iterator == null) {
+            iterator = consumer.poll(ZERO_POLL_TIMEOUT).iterator();
+        }
+        while (iterator.hasNext() && !jobFuture.isDone()) {
+            ConsumerRecord<?, ?> record = iterator.next();
+            Map.Entry<?, ?> entry = entry(record.key(), record.value());
+            // If emit is not successful save the entry as pendingEntry
+            // pause the consumer and do an empty poll for heartbeat
+            if (!tryEmit(entry)) {
+                pendingEntry = entry;
+                pause();
+                emptyPoll();
+                return false;
             }
         }
+        // Records are consumed, commit the offset and reset the iterator for next poll
+        consumer.commitSync();
+        iterator = null;
+        return jobFuture.isDone();
+    }
 
-        return true;
+    private void pause() {
+        consumer.pause(consumer.assignment());
     }
 
-    @Override
-    public boolean isCooperative() {
-        return false;
+    private void resume() {
+        consumer.resume(consumer.assignment());
+    }
+
+    private void emptyPoll() {
```

Comment:
I find this method nome confusing, it makes the same `poll` call as the "nonempty" poll. `pollWhilePaused` would be more self-explanatory.

Also, it uses `poll(0)` while the main path uses `poll(ZERO_POLL_TIMEOUT)`.